src/test/test-openssl.c

```diff
@@ -102,4 +102,56 @@ TEST(invalid) {
         assert_se(pkey == NULL);
 }
 
+static const struct {
+        const char *alg;
+        size_t size;
+} digest_size_table[] = {
+        /* SHA1 "family" */
+        { "sha1",     20, },
+        { "sha-1",    20, },
+        /* SHA2 family */
+        { "sha224",   28, },
+        { "sha-224",  28, },
+        { "sha2-224", 28, },
+        { "sha256",   32, },
+        { "sha-256",  32, },
+        { "sha2-256", 32, },
+        { "sha384",   48, },
+        { "sha-384",  48, },
+        { "sha2-384", 48, },
+        { "sha512",   64, },
+        { "sha-512",  64, },
+        { "sha2-512", 64, },
+        /* SHA3 family */
+        { "sha3-224", 28, },
+        { "sha3-256", 32, },
+        { "sha3-384", 48, },
+        { "sha3-512", 64, },
+        /* SM3 family */
+        { "sm3",      32, },
+        /* MD5 family */
+        { "md5",      16, },
+};
+
+TEST(digest_size) {
+        size_t size;
+        int r;
+
+        FOREACH_ARRAY(t, digest_size_table, ELEMENTSOF(digest_size_table)) {
+                _cleanup_free_ char *uppercase_alg = strdup(t->alg);
+
+                r = openssl_digest_size(t->alg, &size);
+                if (r == -EOPNOTSUPP)
+                        continue;
```

Comment:
would probably be wise to generate some log output here, because it certainly is weird and unexpected if this fails on current openssl versions I guess.

Reply:
Yeah, normally we should expect openssl to support all those algs, but it seems that older openssl (specifically what's in Centos Stream 8) doesn't support the alg naming that includes dashes; or at least, doesn't support some of them. For example:
https://jenkins-systemd.apps.ocp.cloud.ci.centos.org/job/upstream-centos8/5819/artifact/systemd-centos-ci/artifacts_w0ttp9za/testsuite-logs-upstream.AQe/meson-logs/testlog.txt

> /* test_digest_size */
> Digest algorithm 'sha-1' not supported.
> Digest algorithm 'sha-224' not supported.
> Digest algorithm 'sha2-224' not supported.
> Digest algorithm 'sha-256' not supported.
> Digest algorithm 'sha2-256' not supported.
> Digest algorithm 'sha-384' not supported.
> Digest algorithm 'sha2-384' not supported.
> Digest algorithm 'sha-512' not supported.
> Digest algorithm 'sha2-512' not supported.
> Digest algorithm 'invalid.alg' not supported.
> 

It looks like upstream openssl added the names with dashes in this commit from 2019:
https://github.com/openssl/openssl/commit/df553b79419230d698d221919c7ceec68aa8c6c6

I think Centos Stream 8 is getting a bit old...how much longer do we need to stay compatible with it? Is it time to move up to Centos Stream 9? @mrc0mmand?

For this test, it actually might be better to remove the -EOPNOTSUPP check and actually fail if any of these algs aren't supported, and then just skip the test for Centos Stream 8?

Reply:
ah, so it looks like this 2019 commit only made it into openssl 3, so i think the "sha-*" alg names can be ifdef'ed and we can then assume all the algs are supported